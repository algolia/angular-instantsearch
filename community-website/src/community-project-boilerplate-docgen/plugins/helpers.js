const highlight = require('./syntaxHighlighting.js');
const md = require('./mdRenderer.js');

// this plugin provides ATM one helper to easily compute the publicPath of assets
module.exports = function helpers(filenames, metalsmith, cb) {
  // eslint-disable-next-line no-param-reassign
  metalsmith.metadata().h = {
    markdown(src) {
      return md.render(src);
    },
    highlight(src, opts) {
      const lang = opts && opts.lang;
      return highlight(src, lang);
    },
    maybeActive(navPath, singlePathOrArrayOfPaths) {
      const pathsToTest = [].concat(singlePathOrArrayOfPaths);
      return pathsToTest.some(pathToTest => navPath.indexOf(pathToTest) === 0)
        ? 'active'
        : '';
    },
  };

  cb();
};
