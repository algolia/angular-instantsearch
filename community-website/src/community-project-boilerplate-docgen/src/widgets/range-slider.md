---
title: Widgets - Range Slider
layout: widget.pug
canonical: https://www.algolia.com/doc/api-reference/widgets/range-slider/angular/
---

# Range Slider

## Description

The range slider is a widget which provides a user-friendly way to filter the results based on a single numeric range.

You will need to import by yourself the external CSS in order to make the range slider work. The CSS is available:

* `https://cdnjs.cloudflare.com/ajax/libs/noUiSlider/10.1.0/nouislider.min.css` (CDN)
* `node_modules/nouislider/distribute/nouislider.min.css` (LOCAL)

## Options

| Attribute       | Type       | Description
| -                | -          | -
| `attribute`      | `string`   | Name of the attribute for faceting
| `pips?`          | `boolean`  | Show slider pips
| `tooltips?`      | `boolean`  | Should we show tooltips or not.
| `min?`           | `number`   | Minimal slider value, default to automatically computed
| `max?`           | `number`   | Minimal slider value, default to automatically computed

## Code example

```html
<ais-range-slider
  attribute="price"
  [min]="10"
  [max]="20"
>
</ais-range-slider>
```
