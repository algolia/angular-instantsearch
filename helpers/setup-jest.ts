import "jest-preset-angular";

// stupid workaround for TS
import * as helper from "algoliasearch-helper";
require("algoliasearch-helper").default = helper;

const mock = () => {
  let storage = {};
  return {
    getItem(key) {
      return storage[key] || null;
    },
    setItem(key, value) {
      storage[key] = value;
    },
    removeItem(key) {
      delete storage[key];
    },
    clear() {
      storage = {};
    }
  };
};

Object.defineProperty(window, "localStorage", { value: mock() });
Object.defineProperty(window, "sessionStorage", { value: mock() });
Object.defineProperty(window, "getComputedStyle", {
  value: () => ["-webkit-appearance"]
});
