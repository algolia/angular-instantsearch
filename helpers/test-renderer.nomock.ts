import { Component, ViewChild } from '@angular/core';
import { TestBed } from '@angular/core/testing';

import { NgAisInstantSearchModule } from '../src/instantsearch/instantsearch.module';

// same as test-renderer, but without mocking InstantSearch
// In the future, we want to migrate all tests to this

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations,
  methods = {},
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
  methods?: {};
}) {
  return function(state?: {}, isFirstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined,
        methods,
      },
      isFirstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state,
    methods,
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
    methods?: {};
  },
  isFirstRender = false
) {
  @Component({
    template: `
      <ais-instantsearch [config]="config">
        ${template}
      </ais-instantsearch>
    `,
  })
  class TestContainer {
    config = {
      searchClient: {
        search(requests) {
          return Promise.resolve({
            results: requests.map(() => ({ hits: [] })),
          });
        },
      },
      indexName: 'instant_search',
    };
    @ViewChild(TestedWidget) testedWidget;
    constructor() {
      Object.keys(methods).forEach(methodName => {
        this[methodName] = methods[methodName];
      });
    }
  }

  TestBed.configureCompiler({
    preserveWhitespaces: false,
  } as any).configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || []),
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])],
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, isFirstRender);
    fixture.detectChanges();
  }

  return fixture;
}
