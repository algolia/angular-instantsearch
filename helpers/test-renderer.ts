import { Component, ViewChild } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../src/instantsearch/instantsearch.module";

// mock component, dont create a real instantearch instance
jest.mock("../src/base-widget");
jest.mock("../src/instantsearch/instantsearch");

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
}) {
  return function(state?: {}, firstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined
      },
      firstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
  },
  firstRender = false
) {
  @Component({
    template: `
      <ng-ais-instantsearch>
        ${template}
      </ng-ais-instantsearch>
    `
  })
  class TestContainer {
    @ViewChild(TestedWidget) testedWidget;
  }

  TestBed.configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || [])
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])]
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, firstRender);
    fixture.detectChanges();
  }

  return fixture;
}
