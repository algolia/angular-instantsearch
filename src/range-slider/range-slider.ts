import { Component, Input, ViewChild, Inject, forwardRef } from "@angular/core";

import { connectRange } from "instantsearch.js/es/connectors";
import { isPlainObject, noop, omit } from "lodash-es";
import * as noUiSlider from "nouislider";

import { BaseWidget } from "../base-widget";
import { NgAisInstantSearch } from "../instantsearch/instantsearch";
import { parseNumberInput } from "../utils";

export type RangeSliderState = {
  range: { min: number; max: number };
  refine: Function;
  start: number[];
};

@Component({
  selector: "ng-ais-range-slider",
  template: `
    <div [class]="cx()">
      <div [class]="cx('body')">
        <div #sliderContainer></div>
      </div>
    </div>
  `
})
export class NgAisRangeSlider extends BaseWidget {
  @ViewChild("sliderContainer") public sliderContainer: any;

  // render options
  @Input() public pips: boolean = true;
  @Input() public tooltips: boolean = true;

  // connector options
  @Input() public attribute: string;
  @Input() public min?: number | string;
  @Input() public max?: number | string;
  @Input() public precision?: number | string = 2;

  public state: RangeSliderState = {
    range: { min: 0, max: 1 },
    refine: noop,
    start: [0, 1]
  };

  private slider: any;

  get step() {
    // compute step from the precision value
    return 1 / Math.pow(10, parseNumberInput(this.precision));
  }

  constructor(
    @Inject(forwardRef(() => NgAisInstantSearch))
    public instantSearchParent: any
  ) {
    super("RangeSlider");
  }

  public ngOnInit() {
    this.createWidget(connectRange, {
      attributeName: this.attribute,
      max: parseNumberInput(this.max),
      min: parseNumberInput(this.min),
      precision: parseNumberInput(this.precision)
    });

    super.ngOnInit();
  }

  public updateState = (state: RangeSliderState, isFirstRendering: boolean) => {
    if (isFirstRendering) {
      // create slider
      const config = {
        animate: false,
        behaviour: "snap",
        connect: true,
        range: { min: 0, max: 1 },
        start: [0, 1],
        step: this.step,
        tooltips: this.tooltips && [
          { to: this.formatTooltip },
          { to: this.formatTooltip }
        ]
      };

      if (this.pips === true || typeof this.pips === "undefined") {
        Object.assign(config, {
          pips: {
            density: 3,
            mode: "positions",
            stepped: true,
            values: [0, 50, 100]
          }
        });
      } else if (isPlainObject(this.pips)) {
        Object.assign(config, { pips: this.pips });
      }

      this.slider = noUiSlider.create(
        this.sliderContainer.nativeElement,
        config
      );

      // register listen events
      this.sliderContainer.nativeElement.noUiSlider.on(
        "change",
        this.handleChange
      );
    }

    // update component inner state
    this.state = state;

    // update the slider state
    const { range: { min, max }, start } = state;

    const disabled = min === max;
    const range = disabled ? { min, max: max + 0.0001 } : { min, max };

    this.slider.updateOptions({ disabled, range, start });
  };

  public handleChange = (values: string[] | number[]) => {
    this.state.refine(values);
  };

  public formatTooltip = (value: number) => {
    return value.toFixed(parseNumberInput(this.precision));
  };
}
