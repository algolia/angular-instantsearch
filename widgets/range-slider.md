## `<ng-ais-range-slider />`

The range slider is a widget which provides a user-friendly way to filter the results based on a single numeric range.

| Attributes       | Type       | Description
| -                | -          | -
| `attributeName`  | `string`   | Name of the attribute for faceting
| `pips?`          | `boolean`  | Show slider pips
| `tooltips?`      | `boolean`  | Should we show tooltips or not.
| `min?`           | `number`   | Minimal slider value, default to automatically computed
| `max?`           | `number`   | Minimal slider value, default to automatically computed
| `header?`        | `string`   | Displays text before widget
| `footer?`        | `string`   | Displays text after widget