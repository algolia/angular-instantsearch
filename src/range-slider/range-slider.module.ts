import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { NgAisRangeSlider } from './range-slider';

@NgModule({
  declarations: [NgAisRangeSlider],
  entryComponents: [NgAisRangeSlider],
  exports: [NgAisRangeSlider],
  imports: [CommonModule],
})
export class NgAisRangeSliderModule {}
