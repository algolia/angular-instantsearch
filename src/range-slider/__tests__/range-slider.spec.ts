import { createRenderer } from "../../../helpers/test-renderer";
import { NgAisRangeSlider } from "../range-slider";

const defaultState = {
  range: { min: 0, max: 200 },
  refine: jest.fn(),
  start: [0, 200]
};

const render = createRenderer({
  defaultState,
  template: "<ng-ais-range-slider></ng-ais-range-slider>",
  TestedWidget: NgAisRangeSlider
});

describe("RangeSlider", () => {
  it("renders markup without state", () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it("renders markup with state", () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it("should call refine() on slider change", () => {
    const refine = jest.fn();
    const fixture = render({ refine });

    fixture.componentInstance.testedWidget.handleChange([10, 50]);

    expect(refine).toHaveBeenCalled();
    expect(refine).toHaveBeenCalledWith([10, 50]);
  });
});
