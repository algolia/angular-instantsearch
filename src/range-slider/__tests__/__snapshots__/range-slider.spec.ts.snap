// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeSlider renders markup with state 1`] = `
<ng-component
  testedWidget={[Function NgAisRangeSlider]}
>
  <ais-instantsearch>
    <ais-range-slider>
      <div
        class="ais-RangeSlider"
      >
        <div
          class="ais-RangeSlider-body ais-body"
        >
          <div
            class="noUi-target noUi-ltr noUi-horizontal"
          >
            <div
              class="noUi-base"
            >
              <div
                class="noUi-origin"
              >
                <div
                  aria-orientation="horizontal"
                  aria-valuemax="100.0"
                  aria-valuemin="0.0"
                  aria-valuenow="0.0"
                  aria-valuetext="0.00"
                  class="noUi-handle noUi-handle-lower"
                  data-handle="0"
                  role="slider"
                  style="z-index: 5;"
                  tabindex="0"
                >
                  <div
                    class="noUi-tooltip"
                  >
                    0.00
                  </div>
                </div>
              </div>
              <div
                class="noUi-connect"
              />
              <div
                class="noUi-origin"
              >
                <div
                  aria-orientation="horizontal"
                  aria-valuemax="100.0"
                  aria-valuemin="0.0"
                  aria-valuenow="100.0"
                  aria-valuetext="200.00"
                  class="noUi-handle noUi-handle-upper"
                  data-handle="1"
                  role="slider"
                  style="z-index: 4;"
                  tabindex="0"
                >
                  <div
                    class="noUi-tooltip"
                  >
                    200.00
                  </div>
                </div>
              </div>
            </div>
            <div
              class="noUi-pips noUi-pips-horizontal"
            >
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-large"
                style="left: 0.00000%;"
              />
              <div
                class="noUi-value noUi-value-horizontal noUi-value-large"
                style="left: 0.00000%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 2.94118%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 5.88235%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 8.82353%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 11.76471%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 14.70588%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 17.64706%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 20.58824%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 23.52941%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 26.47059%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 29.41176%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 32.35294%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 35.29412%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 38.23529%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 41.17647%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 44.11765%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 47.05882%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-large"
                style="left: 50.00000%;"
              />
              <div
                class="noUi-value noUi-value-horizontal noUi-value-large"
                style="left: 50.00000%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 52.94118%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 55.88235%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 58.82353%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 61.76471%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 64.70588%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 67.64706%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 70.58824%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 73.52941%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 76.47059%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 79.41176%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 82.35294%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 85.29412%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 88.23529%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 91.17647%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 94.11765%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-normal"
                style="left: 97.05882%;"
              />
              <div
                class="noUi-marker noUi-marker-horizontal noUi-marker-large"
                style="left: 100.00000%;"
              />
              <div
                class="noUi-value noUi-value-horizontal noUi-value-large"
                style="left: 100.00000%;"
              />
            </div>
          </div>
        </div>
      </div>
    </ais-range-slider>
  </ais-instantsearch>
</ng-component>
`;

exports[`RangeSlider renders markup without state 1`] = `
<ng-component
  testedWidget={[Function NgAisRangeSlider]}
>
  <ais-instantsearch>
    <ais-range-slider>
      <div
        class="ais-RangeSlider"
      >
        <div
          class="ais-RangeSlider-body ais-body"
        >
          <div />
        </div>
      </div>
    </ais-range-slider>
  </ais-instantsearch>
</ng-component>
`;
